<!DOCTYPE html>
<html lang="en"><head>
<script src="pr_workshop_jan2024_files/libs/clipboard/clipboard.min.js"></script>
<script src="pr_workshop_jan2024_files/libs/quarto-html/tabby.min.js"></script>
<script src="pr_workshop_jan2024_files/libs/quarto-html/popper.min.js"></script>
<script src="pr_workshop_jan2024_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="pr_workshop_jan2024_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="pr_workshop_jan2024_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="pr_workshop_jan2024_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="pr_workshop_jan2024_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.353">

  <meta name="author" content="Phillip Nicol and Rafael Irizarry">
  <title>Workshop on Analyzing Spatial Transcriptomics Data</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="pr_workshop_jan2024_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="pr_workshop_jan2024_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="pr_workshop_jan2024_files/libs/revealjs/dist/theme/quarto.css">
  <link href="pr_workshop_jan2024_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="pr_workshop_jan2024_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="pr_workshop_jan2024_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="pr_workshop_jan2024_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Workshop on Analyzing Spatial Transcriptomics Data</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Phillip Nicol and Rafael Irizarry 
</div>
</div>
</div>

</section>
<section>
<section id="introduction" class="title-slide slide level1 center">
<h1>Introduction</h1>

</section>
<section id="outline" class="slide level2">
<h2>Outline</h2>
</section>
<section id="how-to-follow-along" class="slide level2">
<h2>How to follow along</h2>
</section>
<section id="review-of-spatial-transcriptomics-st-technology" class="slide level2">
<h2>Review of Spatial transcriptomics (ST) technology</h2>
<p>Many of the most popular technologies for ST can be divided into two groups:</p>
<ul>
<li><p><strong>Spot-based:</strong> Next generation sequencing (NGS) with spatial barcoding</p></li>
<li><p><strong>Molecule-based:</strong> In-situ imaging of individual molecules</p></li>
</ul>
<p>FOOTNOTE REF</p>
</section>
<section id="spot-based-10x-genomics-visium" class="slide level2">
<h2>Spot-based: 10X Genomics Visium</h2>
<p><a href="https://www.youtube.com/watch?v=VwNk4d-0RJc">10X Genomics Visium Youtube Video</a></p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="Fig/visium_tissue.png" width="653"></p>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="Fig/visium_barcode.png" width="418"></p>
</figure>
</div>
</section>
<section id="molecule-based-small-molecule-fluorescence-in-situ-hybridization-smfish" class="slide level2 smaller">
<h2>Molecule-based: Small molecule fluorescence in situ hybridization (smFISH)</h2>
<ul>
<li><p>Colored probes attach to mRNA transcript from a target gene.</p></li>
<li><p>Quantify expression by imaging</p></li>
<li><p>Key challenge is extending to many genes</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="Fig/smFISH_example.png" width="248" height="215"></p>
<figcaption>Source: Figure 4A of Ni and Oudenaarden</figcaption>
</figure>
</div>
</section>
<section id="molecule-based-multiplexed-error-robust-fish-merfish" class="slide level2">
<h2>Molecule-based: Multiplexed error robust FISH (MERFISH)</h2>
<p>MERFISH uses (error robust) combinatorial labeling to increase the number of gene transcripts that can be measured.</p>
<p>Basic idea: assign a <span class="math inline">\(N\)</span>-bit binary string to each gene. Then <span class="math inline">\(2^N\)</span> genes can be encoded and measured after <span class="math inline">\(N\)</span> sequential rounds of smFISH.</p>

<img data-src="Fig/merfish.png" width="528" class="r-stretch quarto-figure-center"><p class="caption">Source: Fig 1A of Chen et al.&nbsp;(2015). <em>Science</em>.</p></section>
<section id="comparison" class="slide level2">
<h2>Comparison</h2>
<div style="font-size: 80%;">
<table>
<colgroup>
<col style="width: 26%">
<col style="width: 33%">
<col style="width: 40%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>Visium (Spot-based)</th>
<th>Molecule-based</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Spatial Resolution</strong></td>
<td><span style="color:red;">1-10 cells per spot</span></td>
<td><span style="color:green;">sub-cellular</span></td>
</tr>
<tr class="even">
<td><strong>Number of genes</strong></td>
<td><span style="color:green;">Whole transcriptome (20,000+)</span></td>
<td><span style="color:red;">100-10,000 (MERFISH)</span></td>
</tr>
<tr class="odd">
<td><strong>Detection efficiency</strong></td>
<td><span style="color:red;">~15%</span></td>
<td><span style="color:green;">~95%</span></td>
</tr>
<tr class="even">
<td><strong>Accessibility</strong></td>
<td><span style="color:green;">Commercially available; raw data FASTQ</span></td>
<td><span style="color:red;">Raw images can be large and difficult to analyze</span></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="most-st-studies-choose-visium" class="slide level2">
<h2>Most ST studies choose Visium</h2>

<img data-src="Fig/institutions.png" width="326" class="r-stretch quarto-figure-center"><p class="caption">Moses and Pachter (2022). Nature Methods.</p></section>
<section id="most-st-studies-choose-r" class="slide level2">
<h2>Most ST studies choose R</h2>

<img data-src="Fig/programming_language.png" class="r-stretch quarto-figure-center"><p class="caption">Moses and Pachter (2022). Nature Methods.</p></section></section>
<section>
<section id="spatial-transcriptomics-data-processing" class="title-slide slide level1 center">
<h1>Spatial Transcriptomics Data Processing</h1>

</section>
<section id="upstream-data-processing" class="slide level2">
<h2>Upstream data processing</h2>

<img data-src="Fig/upstream.png" class="r-stretch quarto-figure-center"><p class="caption">Moses and Pachter (2022). Nature Methods.</p></section>
<section id="visium-data-format" class="slide level2">
<h2>Visium data format</h2>
<p>Once processed, ST data can be represented using two matrices.</p>
<ul>
<li><p><strong>Count matrix</strong> <span class="math inline">\(Y\)</span>: For each of <span class="math inline">\(J\)</span> spots record the number of reads from each of <span class="math inline">\(I\)</span> genes.</p></li>
<li><p><strong>Coordinate matrix</strong> <span class="math inline">\(X\)</span>: For each of <span class="math inline">\(J\)</span> spots record the 2-dimensional spatial coordinate.</p></li>
</ul>
<p>FIG</p>
</section>
<section id="storing-st-data-in-r" class="slide level2">
<h2>Storing ST data in R</h2>
<p>The <code>SpatialExperiment</code> package provides a container to store the count and coordinate matrix. ADD REF</p>

<img data-src="Fig/righelli_fig1.png" width="460" class="r-stretch quarto-figure-center"></section></section>
<section>
<section id="creating-a-spatialexperiment-object" class="title-slide slide level1 center">
<h1>Creating a SpatialExperiment object</h1>

</section>
<section id="installing-and-loading-the-spatialexperiment-package" class="slide level2">
<h2>Installing and loading the SpatialExperiment package</h2>
<p>Install the package:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-2"><a href="#cb1-2"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-3"><a href="#cb1-3"></a></span>
<span id="cb1-4"><a href="#cb1-4"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"SpatialExperiment"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Load the package:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="fu">library</span>(SpatialExperiment)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="x-genomics-visium-data" class="slide level2">
<h2>10X genomics Visium data</h2>
<p>The <a href="https://www.10xgenomics.com">10X genomics website</a> has example ST datasets.</p>
<p>For this workshop, we will download samples from human cerebral cortex:</p>

<img data-src="FIg/10xgenomicexample.png" class="r-stretch quarto-figure-center"></section>
<section id="downloading-from-the-website" class="slide level2">
<h2>Downloading from the website</h2>
<p>The data can be downloaded by navigating to the website or directly from the terminal (Unix):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource zsh number-lines code-with-copy"><code class="sourceCode zsh"><span id="cb3-1"><a href="#cb3-1"></a><span class="fu">mkdir</span> cortex</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="bu">cd</span> cortex </span>
<span id="cb3-3"><a href="#cb3-3"></a></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="ex">curl</span> <span class="at">-O</span> https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Human_Brain_Section_1/V1_Human_Brain_Section_1_raw_feature_bc_matrix.tar.gz</span>
<span id="cb3-5"><a href="#cb3-5"></a></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="ex">curl</span> <span class="at">-O</span> https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Human_Brain_Section_1/V1_Human_Brain_Section_1_spatial.tar.gz</span>
<span id="cb3-7"><a href="#cb3-7"></a></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="fu">tar</span> xf V1_Human_Brain_Section_1_raw_feature_bc_matrix.tar.gz</span>
<span id="cb3-9"><a href="#cb3-9"></a></span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="fu">tar</span> xf V1_Human_Brain_Section_1_spatial.tar.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="exercise-creating-the-object" class="slide level2">
<h2>Exercise: Creating the object</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>dir <span class="ot">&lt;-</span> <span class="st">"./cortex"</span></span>
<span id="cb4-2"><a href="#cb4-2"></a></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="do">### Step 1: Load the count matrix</span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="fu">require</span>(DropletUtils)</span>
<span id="cb4-5"><a href="#cb4-5"></a>fnm <span class="ot">&lt;-</span> <span class="fu">file.path</span>(dir, <span class="st">"raw_feature_bc_matrix"</span>)</span>
<span id="cb4-6"><a href="#cb4-6"></a>sce <span class="ot">&lt;-</span> DropletUtils<span class="sc">::</span><span class="fu">read10xCounts</span>(fnm)</span>
<span id="cb4-7"><a href="#cb4-7"></a></span>
<span id="cb4-8"><a href="#cb4-8"></a><span class="do">### Step 2: Load the tissue image</span></span>
<span id="cb4-9"><a href="#cb4-9"></a>img <span class="ot">&lt;-</span> <span class="fu">readImgData</span>(</span>
<span id="cb4-10"><a href="#cb4-10"></a>    <span class="at">path =</span> <span class="fu">file.path</span>(dir, <span class="st">"spatial"</span>),</span>
<span id="cb4-11"><a href="#cb4-11"></a>    <span class="at">sample_id =</span> <span class="st">"brain"</span>)</span>
<span id="cb4-12"><a href="#cb4-12"></a></span>
<span id="cb4-13"><a href="#cb4-13"></a><span class="co"># Step 3: Read spatial coordinates</span></span>
<span id="cb4-14"><a href="#cb4-14"></a>fnm <span class="ot">&lt;-</span> <span class="fu">file.path</span>(dir, <span class="st">"spatial"</span>, <span class="st">"tissue_positions_list.csv"</span>)</span>
<span id="cb4-15"><a href="#cb4-15"></a>xy <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(fnm, <span class="at">header =</span> <span class="cn">FALSE</span>,</span>
<span id="cb4-16"><a href="#cb4-16"></a>    <span class="at">col.names =</span> <span class="fu">c</span>(</span>
<span id="cb4-17"><a href="#cb4-17"></a>        <span class="st">"barcode"</span>, <span class="st">"in_tissue"</span>, <span class="st">"array_row"</span>, <span class="st">"array_col"</span>,</span>
<span id="cb4-18"><a href="#cb4-18"></a>        <span class="st">"pxl_row_in_fullres"</span>, <span class="st">"pxl_col_in_fullres"</span>))</span>
<span id="cb4-19"><a href="#cb4-19"></a>ix <span class="ot">&lt;-</span> <span class="fu">match</span>(sce<span class="sc">$</span>Barcode, xy<span class="sc">$</span>barcode) <span class="co">#Get cells in correct order</span></span>
<span id="cb4-20"><a href="#cb4-20"></a>xy <span class="ot">&lt;-</span> xy[ix,]</span>
<span id="cb4-21"><a href="#cb4-21"></a></span>
<span id="cb4-22"><a href="#cb4-22"></a><span class="do">### Step 4: Construct feature metadata </span></span>
<span id="cb4-23"><a href="#cb4-23"></a><span class="fu">require</span>(S4Vectors)</span>
<span id="cb4-24"><a href="#cb4-24"></a>rd <span class="ot">&lt;-</span> S4Vectors<span class="sc">::</span><span class="fu">DataFrame</span>(</span>
<span id="cb4-25"><a href="#cb4-25"></a>    <span class="at">symbol =</span> <span class="fu">rowData</span>(sce)<span class="sc">$</span>Symbol)</span>
<span id="cb4-26"><a href="#cb4-26"></a></span>
<span id="cb4-27"><a href="#cb4-27"></a><span class="do">### Step 5: Create the object</span></span>
<span id="cb4-28"><a href="#cb4-28"></a>spe <span class="ot">&lt;-</span> <span class="fu">SpatialExperiment</span>(</span>
<span id="cb4-29"><a href="#cb4-29"></a>    <span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> <span class="fu">assay</span>(sce)),</span>
<span id="cb4-30"><a href="#cb4-30"></a>    <span class="at">rowData =</span> rd, </span>
<span id="cb4-31"><a href="#cb4-31"></a>    <span class="at">colData =</span> <span class="fu">DataFrame</span>(xy), </span>
<span id="cb4-32"><a href="#cb4-32"></a>    <span class="at">spatialCoordsNames =</span> <span class="fu">c</span>(<span class="st">"pxl_col_in_fullres"</span>, <span class="st">"pxl_row_in_fullres"</span>),</span>
<span id="cb4-33"><a href="#cb4-33"></a>    <span class="at">imgData =</span> img,</span>
<span id="cb4-34"><a href="#cb4-34"></a>    <span class="at">sample_id =</span> <span class="st">"brain"</span>)</span>
<span id="cb4-35"><a href="#cb4-35"></a></span>
<span id="cb4-36"><a href="#cb4-36"></a><span class="do">### Step 6: Save the object </span></span>
<span id="cb4-37"><a href="#cb4-37"></a><span class="fu">saveRDS</span>(spe, <span class="at">file=</span><span class="fu">file.path</span>(dir,<span class="st">"spe.RDS"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plotting-an-image-of-the-tissue" class="slide level2" data-background-color="black">
<h2>Plotting an image of the tissue</h2>
<div class="panel-tabset">
<ul id="tabset-1" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-1-1">Plot</a></li><li><a href="#tabset-1-2">Code</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="pr_workshop_jan2024_files/figure-revealjs/unnamed-chunk-7-1.png" width="960"></p>
</div>
</div>
</div>
<div id="tabset-1-2">
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>spi <span class="ot">&lt;-</span> <span class="fu">getImg</span>(spe)</span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="fu">plot</span>(<span class="fu">as.raster</span>(spi))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="removing-spots-not-in-the-tissue" class="slide level2">
<h2>Removing spots not in the tissue</h2>
<div class="panel-tabset">
<ul id="tabset-2" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-2-1">Plot</a></li><li><a href="#tabset-2-2">Code</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="pr_workshop_jan2024_files/figure-revealjs/unnamed-chunk-9-1.png" width="960"></p>
</div>
</div>
</div>
<div id="tabset-2-2">
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="fu">require</span>(ggplot2)</span>
<span id="cb6-2"><a href="#cb6-2"></a></span>
<span id="cb6-3"><a href="#cb6-3"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">x=</span><span class="fu">spatialCoords</span>(spe)[,<span class="dv">1</span>], <span class="at">y=</span><span class="fu">spatialCoords</span>(spe)[,<span class="dv">2</span>],</span>
<span id="cb6-4"><a href="#cb6-4"></a>                 <span class="at">in_tissue=</span><span class="fu">as.character</span>(spe<span class="sc">$</span>in_tissue))</span>
<span id="cb6-5"><a href="#cb6-5"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df,<span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">y=</span>y,<span class="at">color=</span>in_tissue))<span class="sc">+</span></span>
<span id="cb6-6"><a href="#cb6-6"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.25</span>)<span class="sc">+</span></span>
<span id="cb6-7"><a href="#cb6-7"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"grey"</span>, <span class="st">"red"</span>))<span class="sc">+</span></span>
<span id="cb6-8"><a href="#cb6-8"></a>  <span class="fu">theme_bw</span>()</span>
<span id="cb6-9"><a href="#cb6-9"></a>p</span>
<span id="cb6-10"><a href="#cb6-10"></a></span>
<span id="cb6-11"><a href="#cb6-11"></a>spe <span class="ot">&lt;-</span> spe[,spe<span class="sc">$</span>in_tissue <span class="sc">==</span> <span class="dv">1</span>] <span class="co">#This is how you subset by spots/samples</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="more-example-st-data" class="slide level2">
<h2>More example ST data</h2>
<p>For additional examples of data in the <code>SpatialExperiment</code> format, a good package is <code>STexampleData</code> :</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"STexampleData"</span>)</span>
<span id="cb7-2"><a href="#cb7-2"></a></span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="do">### Example dataset </span></span>
<span id="cb7-4"><a href="#cb7-4"></a>spe2 <span class="ot">&lt;-</span> STexampleData<span class="sc">::</span><span class="fu">ST_mouseOB</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section>
<section>
<section id="finding-spatially-variable-genes" class="title-slide slide level1 center">
<h1>Finding Spatially Variable Genes</h1>

</section>
<section id="what-is-a-spatially-variable-gene-svg" class="slide level2">
<h2>What is a spatially variable gene (SVG)?</h2>
<p>A spatially variable gene is one whose expression depends on spatial location</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="Fig/SVG_example.png" width="718" height="202"></p>
<figcaption>Source: REF</figcaption>
</figure>
</div>
</section>
<section id="exercise-plotting-spatially-variable-genes" class="slide level2">
<h2>Exercise: Plotting spatially variable genes</h2>
<p>In the human cerebral cortex, the genes <em>MOBP, NPY, SNAP25,</em> and <em>PCP4</em> were previously reported by REF to be spatially variable. We will verify this by plotting.</p>
<p><strong>Problem:</strong> Instead of gene symbols we are given ENSEMBL ID:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="fu">head</span>(<span class="fu">rownames</span>(spe))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ENSG00000243485" "ENSG00000237613" "ENSG00000186092" "ENSG00000238009"
[5] "ENSG00000239945" "ENSG00000239906"</code></pre>
</div>
</div>
</section>
<section id="exercise-plotting-svgs" class="slide level2">
<h2>Exercise: Plotting SVGs</h2>
<div class="panel-tabset">
<ul id="tabset-3" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-3-1">Step 1</a></li><li><a href="#tabset-3-2">Step 2</a></li><li><a href="#tabset-3-3">Step 3</a></li><li><a href="#tabset-3-4">Final Plot</a></li></ul>
<div class="tab-content">
<div id="tabset-3-1">
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="do">### Convert ENSEMBL ID to gene symbol</span></span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="fu">require</span>(biomaRt)</span>
<span id="cb10-3"><a href="#cb10-3"></a>mart <span class="ot">&lt;-</span> <span class="fu">useDataset</span>(<span class="st">"hsapiens_gene_ensembl"</span>, <span class="fu">useMart</span>(<span class="st">"ensembl"</span>))</span>
<span id="cb10-4"><a href="#cb10-4"></a>genes <span class="ot">&lt;-</span> <span class="fu">rownames</span>(spe)</span>
<span id="cb10-5"><a href="#cb10-5"></a>G_list <span class="ot">&lt;-</span> <span class="fu">getBM</span>(<span class="at">filters=</span><span class="st">"ensembl_gene_id"</span>,</span>
<span id="cb10-6"><a href="#cb10-6"></a>                <span class="at">attributes=</span> <span class="fu">c</span>(<span class="st">"ensembl_gene_id"</span>,<span class="st">"hgnc_symbol"</span>),</span>
<span id="cb10-7"><a href="#cb10-7"></a>                <span class="at">values=</span>genes,</span>
<span id="cb10-8"><a href="#cb10-8"></a>                <span class="at">mart=</span>mart)</span>
<span id="cb10-9"><a href="#cb10-9"></a>no.match <span class="ot">&lt;-</span> <span class="fu">which</span>(G_list<span class="sc">$</span>hgnc_symbol <span class="sc">==</span> <span class="st">""</span>)</span>
<span id="cb10-10"><a href="#cb10-10"></a>G_list[no.match,<span class="dv">2</span>] <span class="ot">&lt;-</span> G_list[no.match,<span class="dv">1</span>]</span>
<span id="cb10-11"><a href="#cb10-11"></a>ix <span class="ot">&lt;-</span> <span class="fu">match</span>(G_list<span class="sc">$</span>ensembl_gene_id, <span class="fu">rownames</span>(spe))</span>
<span id="cb10-12"><a href="#cb10-12"></a><span class="fu">rownames</span>(spe)[ix] <span class="ot">&lt;-</span> G_list<span class="sc">$</span>hgnc_symbol</span>
<span id="cb10-13"><a href="#cb10-13"></a>interesting_genes <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">rownames</span>(spe) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"MOBP"</span>,<span class="st">"PCP4"</span>, <span class="st">"SNAP25"</span>,<span class="st">"NPY"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-3-2">
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="do">### Normalize count matrix and subset to interesting genes</span></span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="fu">require</span>(scran)</span>
<span id="cb11-3"><a href="#cb11-3"></a>spe <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(spe)</span>
<span id="cb11-4"><a href="#cb11-4"></a></span>
<span id="cb11-5"><a href="#cb11-5"></a>Y.sub <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">logcounts</span>(spe)[interesting_genes,])</span>
<span id="cb11-6"><a href="#cb11-6"></a></span>
<span id="cb11-7"><a href="#cb11-7"></a><span class="do">## Switch rows and columns and make dataframe </span></span>
<span id="cb11-8"><a href="#cb11-8"></a>df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(Y.sub))</span>
<span id="cb11-9"><a href="#cb11-9"></a></span>
<span id="cb11-10"><a href="#cb11-10"></a><span class="do">## Add spatial coordinates </span></span>
<span id="cb11-11"><a href="#cb11-11"></a>df<span class="sc">$</span>x <span class="ot">&lt;-</span> <span class="fu">spatialCoords</span>(spe)[,<span class="dv">1</span>]; df<span class="sc">$</span>y <span class="ot">&lt;-</span> <span class="fu">spatialCoords</span>(spe)[,<span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-3-3">
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="do">### Construct plot</span></span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="fu">require</span>(reshape2)</span>
<span id="cb12-3"><a href="#cb12-3"></a>df <span class="ot">&lt;-</span> <span class="fu">melt</span>(df, <span class="at">id.vars=</span><span class="fu">c</span>(<span class="st">"x"</span>,<span class="st">"y"</span>))</span>
<span id="cb12-4"><a href="#cb12-4"></a></span>
<span id="cb12-5"><a href="#cb12-5"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>df,<span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">y=</span>y,<span class="at">color=</span>value)) </span>
<span id="cb12-6"><a href="#cb12-6"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.05</span>) </span>
<span id="cb12-7"><a href="#cb12-7"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span>
<span id="cb12-8"><a href="#cb12-8"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_color_gradientn</span>(<span class="at">colors=</span><span class="fu">c</span>(<span class="st">"gray90"</span>,<span class="st">"blue"</span>,<span class="st">"black"</span>),</span>
<span id="cb12-9"><a href="#cb12-9"></a>                              <span class="at">breaks=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">3</span>,<span class="dv">6</span>))</span>
<span id="cb12-10"><a href="#cb12-10"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>variable, <span class="at">nrow=</span><span class="dv">2</span>)</span>
<span id="cb12-11"><a href="#cb12-11"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_bw</span>()</span>
<span id="cb12-12"><a href="#cb12-12"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_colorbar</span>(<span class="at">ticks =</span> <span class="cn">FALSE</span>)) </span>
<span id="cb12-13"><a href="#cb12-13"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">color=</span><span class="st">"log count"</span>)</span>
<span id="cb12-14"><a href="#cb12-14"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">strip.text =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">"italic"</span>), </span>
<span id="cb12-15"><a href="#cb12-15"></a>        <span class="at">panel.grid =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb12-16"><a href="#cb12-16"></a>        <span class="at">axis.title =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb12-17"><a href="#cb12-17"></a>        <span class="at">axis.text =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb12-18"><a href="#cb12-18"></a>        <span class="at">axis.ticks =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-3-4">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="pr_workshop_jan2024_files/figure-revealjs/unnamed-chunk-17-1.png" width="960"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="finding-svgs-automatically" class="slide level2">
<h2>Finding SVGs automatically</h2>
<ul>
<li><p>A statistical method can automatically identify spatially variable genes without relying on previous knowledge.</p></li>
<li><p>There are several existing methods, but it is still an area of active research.</p></li>
</ul>
</section>
<section id="spark-x" class="slide level2">
<h2>SPARK-X</h2>
<p>METHOD OVERVIEW</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="fu">install.packages</span>(<span class="st">'devtools'</span>) <span class="co">#If not already installed</span></span>
<span id="cb13-2"><a href="#cb13-2"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">'xzhoulab/SPARK'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="running-spark-x" class="slide level2">
<h2>Running SPARK-X</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="fu">require</span>(SPARK)</span>
<span id="cb14-2"><a href="#cb14-2"></a></span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="do">## Remove mitochondrial genes</span></span>
<span id="cb14-4"><a href="#cb14-4"></a>mito.genes <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">grepl</span>(<span class="st">"^MT-"</span>,<span class="fu">rownames</span>(spe)))</span>
<span id="cb14-5"><a href="#cb14-5"></a>spe <span class="ot">&lt;-</span> spe[<span class="sc">-</span>mito.genes,]</span>
<span id="cb14-6"><a href="#cb14-6"></a></span>
<span id="cb14-7"><a href="#cb14-7"></a>res <span class="ot">&lt;-</span> <span class="fu">sparkx</span>(</span>
<span id="cb14-8"><a href="#cb14-8"></a>    <span class="at">count_in =</span> <span class="fu">counts</span>(spe)[<span class="fu">rowSums</span>(<span class="fu">counts</span>(spe)) <span class="sc">&gt;</span> <span class="dv">10</span>,], </span>
<span id="cb14-9"><a href="#cb14-9"></a>    <span class="at">locus_in =</span> <span class="fu">spatialCoords</span>(spe), </span>
<span id="cb14-10"><a href="#cb14-10"></a>    <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="genes-with-the-smallest-p-values" class="slide level2">
<h2>Genes with the smallest p-values</h2>
<div class="panel-tabset">
<ul id="tabset-4" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-4-1">Plot</a></li><li><a href="#tabset-4-2">Code</a></li></ul>
<div class="tab-content">
<div id="tabset-4-1">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="pr_workshop_jan2024_files/figure-revealjs/unnamed-chunk-21-1.png" width="960"></p>
</div>
</div>
</div>
<div id="tabset-4-2">
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="fu">require</span>(scran)</span>
<span id="cb15-2"><a href="#cb15-2"></a>interesting_genes <span class="ot">&lt;-</span> <span class="fu">rownames</span>(res<span class="sc">$</span>res_mtest)[<span class="fu">order</span>(res<span class="sc">$</span>res_mtest<span class="sc">$</span>combinedPval)[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]]</span>
<span id="cb15-3"><a href="#cb15-3"></a></span>
<span id="cb15-4"><a href="#cb15-4"></a>spe <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(spe)</span>
<span id="cb15-5"><a href="#cb15-5"></a></span>
<span id="cb15-6"><a href="#cb15-6"></a>Y.sub <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">logcounts</span>(spe)[interesting_genes,])</span>
<span id="cb15-7"><a href="#cb15-7"></a></span>
<span id="cb15-8"><a href="#cb15-8"></a><span class="do">## Switch rows and columns and make dataframe </span></span>
<span id="cb15-9"><a href="#cb15-9"></a>df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(Y.sub))</span>
<span id="cb15-10"><a href="#cb15-10"></a></span>
<span id="cb15-11"><a href="#cb15-11"></a><span class="do">## Add spatial coordinates </span></span>
<span id="cb15-12"><a href="#cb15-12"></a>df<span class="sc">$</span>x <span class="ot">&lt;-</span> <span class="fu">spatialCoords</span>(spe)[,<span class="dv">1</span>]; df<span class="sc">$</span>y <span class="ot">&lt;-</span> <span class="fu">spatialCoords</span>(spe)[,<span class="dv">2</span>]</span>
<span id="cb15-13"><a href="#cb15-13"></a></span>
<span id="cb15-14"><a href="#cb15-14"></a><span class="do">### Construct plot</span></span>
<span id="cb15-15"><a href="#cb15-15"></a><span class="fu">require</span>(reshape2)</span>
<span id="cb15-16"><a href="#cb15-16"></a>df <span class="ot">&lt;-</span> <span class="fu">melt</span>(df, <span class="at">id.vars=</span><span class="fu">c</span>(<span class="st">"x"</span>,<span class="st">"y"</span>))</span>
<span id="cb15-17"><a href="#cb15-17"></a></span>
<span id="cb15-18"><a href="#cb15-18"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data=</span>df,<span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">y=</span>y,<span class="at">color=</span>value)) </span>
<span id="cb15-19"><a href="#cb15-19"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.05</span>) </span>
<span id="cb15-20"><a href="#cb15-20"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span>
<span id="cb15-21"><a href="#cb15-21"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">scale_color_gradientn</span>(<span class="at">colors=</span><span class="fu">c</span>(<span class="st">"gray90"</span>,<span class="st">"blue"</span>,<span class="st">"black"</span>),</span>
<span id="cb15-22"><a href="#cb15-22"></a>                              <span class="at">breaks=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">4</span>,<span class="dv">8</span>))</span>
<span id="cb15-23"><a href="#cb15-23"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>variable, <span class="at">nrow=</span><span class="dv">2</span>)</span>
<span id="cb15-24"><a href="#cb15-24"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme_bw</span>()</span>
<span id="cb15-25"><a href="#cb15-25"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_colorbar</span>(<span class="at">ticks =</span> <span class="cn">FALSE</span>)) </span>
<span id="cb15-26"><a href="#cb15-26"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">color=</span><span class="st">"log count"</span>)</span>
<span id="cb15-27"><a href="#cb15-27"></a>p <span class="ot">&lt;-</span> p <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">strip.text =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">"italic"</span>), </span>
<span id="cb15-28"><a href="#cb15-28"></a>        <span class="at">panel.grid =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb15-29"><a href="#cb15-29"></a>        <span class="at">axis.title =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb15-30"><a href="#cb15-30"></a>        <span class="at">axis.text =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb15-31"><a href="#cb15-31"></a>        <span class="at">axis.ticks =</span> <span class="fu">element_blank</span>())</span>
<span id="cb15-32"><a href="#cb15-32"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="plotting-genes-with-the-largest-p-values-least-spatially-variable" class="slide level2">
<h2>Plotting genes with the largest p-values (least spatially variable)</h2>

<img data-src="pr_workshop_jan2024_files/figure-revealjs/unnamed-chunk-23-1.png" width="960" class="r-stretch"></section></section>
<section>
<section id="clustering-and-cell-type-identification" class="title-slide slide level1 center">
<h1>Clustering and cell type identification</h1>

</section>
<section id="dimension-reduction" class="slide level2">
<h2>Dimension reduction</h2>
</section>
<section id="running-pca" class="slide level2">
<h2>Running PCA</h2>
</section>
<section id="plotting-top-3-pcs" class="slide level2">
<h2>Plotting top 3 PCs</h2>

<img data-src="pr_workshop_jan2024_files/figure-revealjs/unnamed-chunk-25-1.png" width="960" class="r-stretch"></section>
<section id="clustering" class="slide level2">
<h2>Clustering</h2>

<img data-src="pr_workshop_jan2024_files/figure-revealjs/unnamed-chunk-27-1.png" width="960" class="r-stretch"></section>
<section id="cell-type-identification" class="slide level2">
<h2>Cell type identification</h2>
</section>
<section id="references" class="slide level2">
<h2>References</h2>
<p>Ji, N., &amp; Van Oudenaarden, A. (2012). Single molecule fluorescent in situ hybridization (smFISH) of C. elegans worms and embryos.</p>
<div class="footer footer-default">

</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="pr_workshop_jan2024_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="pr_workshop_jan2024_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="pr_workshop_jan2024_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="pr_workshop_jan2024_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="pr_workshop_jan2024_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="pr_workshop_jan2024_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="pr_workshop_jan2024_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="pr_workshop_jan2024_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="pr_workshop_jan2024_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="pr_workshop_jan2024_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: false,

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>