---
title: "Workshop on Analyzing Spatial Transcriptomics Data"
author: "Phillip Nicol and Rafael Irizarry"
format: revealjs
editor: visual
---

# Introduction

## Outline

## How to follow along

```{r}
require(tidyverse)
```

## Review of Spatial transcriptomics (ST) technology

Many of the most popular technologies for ST can be divided into two groups:

-   **Spot-based:** Next generation sequencing (NGS) with spatial barcoding

-   **Molecule-based:** In-situ imaging of individual molecules

FOOTNOTE REF

## Spot-based: 10X Genomics Visium

[10X Genomics Visium Youtube Video](https://www.youtube.com/watch?v=VwNk4d-0RJc)

![](Fig/visium_tissue.png){fig-align="center" width="653"}

![](Fig/visium_barcode.png){fig-align="center" width="418"}

## Molecule-based: Small molecule fluorescence in situ hybridization (smFISH) {.smaller}

-   Colored probes attach to mRNA transcript from a target gene.

-   Quantify expression by imaging

-   Key challenge is extending to many genes

![Source: Figure 4A of Ni and Oudenaarden](Fig/smFISH_example.png){fig-align="center" width="248" height="215"}

## Molecule-based: Multiplexed error robust FISH (MERFISH)

MERFISH uses (error robust) combinatorial labeling to increase the number of gene transcripts that can be measured.

Basic idea: assign a $N$-bit binary string to each gene. Then $2^N$ genes can be encoded and measured after $N$ sequential rounds of smFISH.

![Source: Fig 1A of Chen et al. (2015). *Science*.](Fig/merfish.png){fig-align="center" width="528"}

## Comparison

::: {style="font-size: 80%;"}
|                          | Visium (Spot-based)                                            | Molecule-based                                                         |
|-------------------|------------------------|-----------------------------|
| **Spatial Resolution**   | [1-10 cells per spot]{style="color:red;"}                      | [sub-cellular]{style="color:green;"}                                   |
| **Number of genes**      | [Whole transcriptome (20,000+)]{style="color:green;"}          | [100-10,000 (MERFISH)]{style="color:red;"}                             |
| **Detection efficiency** | [\~15%]{style="color:red;"}                                    | [\~95%]{style="color:green;"}                                          |
| **Accessibility**        | [Commercially available; raw data FASTQ]{style="color:green;"} | [Raw images can be large and difficult to analyze]{style="color:red;"} |
:::

## Most ST studies choose Visium

![Moses and Pachter (2022). Nature Methods.](Fig/institutions.png){fig-align="center" width="326"}

## Most ST studies choose R

![Moses and Pachter (2022). Nature Methods.](Fig/programming_language.png){fig-align="center"}

# Spatial Transcriptomics Data Processing

## Upstream data processing

![Moses and Pachter (2022). Nature Methods.](Fig/upstream.png){fig-align="center"}

## Visium data format

Once processed, ST data can be represented using two matrices.

-   **Count matrix** $Y$: For each of $J$ spots record the number of reads from each of $I$ genes.

-   **Coordinate matrix** $X$: For each of $J$ spots record the 2-dimensional spatial coordinate.

FIG

## Storing ST data in R

The `SpatialExperiment` package provides a container to store the count and coordinate matrix. ADD REF

![](Fig/righelli_fig1.png){fig-align="center" width="460"}

# Creating a SpatialExperiment object

## Installing and loading the SpatialExperiment package

Install the package:

```{r, eval=FALSE, echo=TRUE}

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("SpatialExperiment")
```

Load the package:

```{r, echo=TRUE}
library(SpatialExperiment)
```

## 10X genomics Visium data

The [10X genomics website](https://www.10xgenomics.com) has example ST datasets.

For this workshop, we will download samples from human cerebral cortex:

![](FIg/10xgenomicexample.png){fig-align="center"}

## Downloading from the website

The data can be downloaded by navigating to the website or directly from the terminal (Unix):

```{zsh, echo=TRUE, eval=FALSE}
mkdir cortex
cd cortex 

curl -O https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Human_Brain_Section_1/V1_Human_Brain_Section_1_raw_feature_bc_matrix.tar.gz

curl -O https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Human_Brain_Section_1/V1_Human_Brain_Section_1_spatial.tar.gz

tar xf V1_Human_Brain_Section_1_raw_feature_bc_matrix.tar.gz

tar xf V1_Human_Brain_Section_1_spatial.tar.gz
```

## Exercise: Creating the object

```{r, echo=TRUE, eval=FALSE}
dir <- "./cortex"

### Step 1: Load the count matrix
require(DropletUtils)
fnm <- file.path(dir, "raw_feature_bc_matrix")
sce <- DropletUtils::read10xCounts(fnm)

### Step 2: Load the tissue image
img <- readImgData(
    path = file.path(dir, "spatial"),
    sample_id = "brain")

# Step 3: Read spatial coordinates
fnm <- file.path(dir, "spatial", "tissue_positions_list.csv")
xy <- read.csv(fnm, header = FALSE,
    col.names = c(
        "barcode", "in_tissue", "array_row", "array_col",
        "pxl_row_in_fullres", "pxl_col_in_fullres"))
ix <- match(sce$Barcode, xy$barcode) #Get cells in correct order
xy <- xy[ix,]

### Step 4: Construct feature metadata 
require(S4Vectors)
rd <- S4Vectors::DataFrame(
    symbol = rowData(sce)$Symbol)

### Step 5: Create the object
spe <- SpatialExperiment(
    assays = list(counts = assay(sce)),
    rowData = rd, 
    colData = DataFrame(xy), 
    spatialCoordsNames = c("pxl_col_in_fullres", "pxl_row_in_fullres"),
    imgData = img,
    sample_id = "brain")

### Step 6: Save the object 
saveRDS(spe, file=file.path(dir,"spe.RDS"))
```

## Plotting an image of the tissue {background-color="black"}

```{r}
spe <- readRDS("./cortex/spe.RDS")
```

::: panel-tabset
## Plot

```{r}
spi <- getImg(spe)
par(bg = 'black', mar=c(0,0,0,0))
plot(as.raster(spi))
```

## Code

```{r, echo=TRUE, eval=FALSE}
spi <- getImg(spe)
plot(as.raster(spi))
```
:::

## Removing spots not in the tissue

::: panel-tabset
## Plot

```{r}
require(ggplot2)
in_tissue <- spe$in_tissue 
spe <- spe[,spe$in_tissue == 1] #This is how you subset by spots/samples

df <- data.frame(x=spatialCoords(spe)[,1], y=spatialCoords(spe)[,2])
p <- ggplot(df,aes(x=x,y=y)) + geom_point(size=0.05)
p <- p + theme_bw()
p
```

## Code

```{r, echo=TRUE, eval=FALSE}
require(ggplot2)

df <- data.frame(x=spatialCoords(spe)[,1], y=spatialCoords(spe)[,2],
                 in_tissue=as.character(spe$in_tissue))
p <- ggplot(df,aes(x=x,y=y,color=in_tissue))+
  geom_point(size=0.25)+
  scale_color_manual(values=c("grey", "red"))+
  theme_bw()
p

spe <- spe[,spe$in_tissue == 1] #This is how you subset by spots/samples
```
:::

## More example ST data

For additional examples of data in the `SpatialExperiment` format, a good package is `STexampleData` :

```{r, echo=TRUE, eval=FALSE}
BiocManager::install("STexampleData")

### Example dataset 
spe2 <- STexampleData::ST_mouseOB()
```

# Finding Spatially Variable Genes

## What is a spatially variable gene (SVG)?

A spatially variable gene is one whose expression depends on spatial location

![Source: REF](Fig/SVG_example.png){fig-align="center" width="718" height="202"}

```{r, eval=FALSE}
library(biomaRt)
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
genes <- rownames(spe)
G_list <- getBM(filters="ensembl_gene_id",
                attributes= c("ensembl_gene_id","hgnc_symbol"),
                values=genes,
                mart=mart)
no.match <- which(G_list$hgnc_symbol == "")
G_list[no.match,2] <- G_list[no.match,1]
ix <- match(G_list$ensembl_gene_id, rownames(spe))
rownames(spe)[ix] <- G_list$hgnc_symbol
interesting_genes <- which(rownames(spe) %in% c("MOBP","PCP4", "SNAP25","NPY"))


library(scran)
spe <- logNormCounts(spe)

Y.sub <- as.matrix(logcounts(spe)[interesting_genes,])

## Switch rows and columns and make dataframe 
df <- as.data.frame(t(Y.sub))

## Add spatial coordinates 
df$x <- spatialCoords(spe)[,1]; df$y <- spatialCoords(spe)[,2]

library(reshape2)
df <- melt(df, id.vars=c("x","y"))
saveRDS(df, file="./analysis/svgplot_df.RDS")
saveRDS(spe, file=file.path("./cortex/spe_p2.RDS"))
```

## Exercise: Plotting spatially variable genes

In the human cerebral cortex, the genes *MOBP, NPY, SNAP25,* and *PCP4* were previously reported by REF to be spatially variable. We will verify this by plotting.

**Problem:** Instead of gene symbols we are given ENSEMBL ID:

```{r, echo=TRUE}
head(rownames(spe))
```

## Exercise: Plotting SVGs

::: panel-tabset
## Step 1

```{r, echo=TRUE, eval=FALSE}
### Convert ENSEMBL ID to gene symbol
require(biomaRt)
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
genes <- rownames(spe)
G_list <- getBM(filters="ensembl_gene_id",
                attributes= c("ensembl_gene_id","hgnc_symbol"),
                values=genes,
                mart=mart)
no.match <- which(G_list$hgnc_symbol == "")
G_list[no.match,2] <- G_list[no.match,1]
ix <- match(G_list$ensembl_gene_id, rownames(spe))
rownames(spe)[ix] <- G_list$hgnc_symbol
interesting_genes <- which(rownames(spe) %in% c("MOBP","PCP4", "SNAP25","NPY"))
```

## Step 2

```{r, echo=TRUE, eval=FALSE}
### Normalize count matrix and subset to interesting genes
require(scran)
spe <- logNormCounts(spe)

Y.sub <- as.matrix(logcounts(spe)[interesting_genes,])

## Switch rows and columns and make dataframe 
df <- as.data.frame(t(Y.sub))

## Add spatial coordinates 
df$x <- spatialCoords(spe)[,1]; df$y <- spatialCoords(spe)[,2]
```

## Step 3

```{r, echo=TRUE, eval=FALSE}
### Construct plot
require(reshape2)
df <- melt(df, id.vars=c("x","y"))

p <- ggplot(data=df,aes(x=x,y=y,color=value)) 
p <- p + geom_point(size=0.05) 
p <- p + coord_fixed()
p <- p + scale_color_gradientn(colors=c("gray90","blue","black"),
                              breaks=c(0,3,6))
p <- p + facet_wrap(~variable, nrow=2)
p <- p + theme_bw()
p <- p + guides(color = guide_colorbar(ticks = FALSE)) 
p <- p + labs(color="log count")
p <- p + theme(strip.text = element_text(face = "italic"), 
        panel.grid = element_blank(), 
        axis.title = element_blank(), 
        axis.text = element_blank(), 
        axis.ticks = element_blank())
```

## Final Plot

```{r, echo=FALSE}
df <- readRDS(file="./analysis/svgplot_df.RDS")

p <- ggplot(data=df,aes(x=x,y=y,color=value)) 
p <- p + geom_point(size=0.05) 
p <- p + coord_fixed()
p <- p + scale_color_gradientn(colors=c("gray90","blue","black"),
                              breaks=c(0,3,6))
p <- p + facet_wrap(~variable, nrow=2)
p <- p + theme_bw()
p <- p + guides(color = guide_colorbar(ticks = FALSE)) 
p <- p + labs(color="log count")
p <- p + theme(strip.text = element_text(face = "italic"), 
        panel.grid = element_blank(), 
        axis.title = element_blank(), 
        axis.text = element_blank(), 
        axis.ticks = element_blank())
p
```
:::

## Finding SVGs automatically

-   A statistical method can automatically identify spatially variable genes without relying on previous knowledge.

-   There are several existing methods, but it is still an area of active research.

## SPARK-X 

METHOD OVERVIEW

```{r, echo=TRUE, eval=FALSE}
install.packages('devtools') #If not already installed
devtools::install_github('xzhoulab/SPARK')
```

```{r}
spe <- readRDS("./cortex/spe_p2.RDS")
```

## Running SPARK-X

```{r, echo=TRUE, output=FALSE}
require(SPARK)

## Remove mitochondrial genes
mito.genes <- which(grepl("^MT-",rownames(spe)))
spe <- spe[-mito.genes,]

res <- sparkx(
    count_in = counts(spe)[rowSums(counts(spe)) > 10,], 
    locus_in = spatialCoords(spe), 
    verbose = FALSE)
```

## Genes with the smallest p-values

::: panel-tabset
## Plot

```{r}
require(scran)
interesting_genes <- rownames(res$res_mtest)[order(res$res_mtest$combinedPval)[1:10]]

spe <- logNormCounts(spe)

Y.sub <- as.matrix(logcounts(spe)[interesting_genes,])

## Switch rows and columns and make dataframe 
df <- as.data.frame(t(Y.sub))

## Add spatial coordinates 
df$x <- spatialCoords(spe)[,1]; df$y <- spatialCoords(spe)[,2]

### Construct plot
require(reshape2)
df <- melt(df, id.vars=c("x","y"))

p <- ggplot(data=df,aes(x=x,y=y,color=value)) 
p <- p + geom_point(size=0.05) 
p <- p + coord_fixed()
p <- p + scale_color_gradientn(colors=c("gray90","blue","black"),
                              breaks=c(0,4,8))
p <- p + facet_wrap(~variable, nrow=2)
p <- p + theme_bw()
p <- p + guides(color = guide_colorbar(ticks = FALSE)) 
p <- p + labs(color="log count")
p <- p + theme(strip.text = element_text(face = "italic"), 
        panel.grid = element_blank(), 
        axis.title = element_blank(), 
        axis.text = element_blank(), 
        axis.ticks = element_blank())
p
```

## Code

```{r, eval=FALSE, echo=TRUE}
require(scran)
interesting_genes <- rownames(res$res_mtest)[order(res$res_mtest$combinedPval)[1:10]]

spe <- logNormCounts(spe)

Y.sub <- as.matrix(logcounts(spe)[interesting_genes,])

## Switch rows and columns and make dataframe 
df <- as.data.frame(t(Y.sub))

## Add spatial coordinates 
df$x <- spatialCoords(spe)[,1]; df$y <- spatialCoords(spe)[,2]

### Construct plot
require(reshape2)
df <- melt(df, id.vars=c("x","y"))

p <- ggplot(data=df,aes(x=x,y=y,color=value)) 
p <- p + geom_point(size=0.05) 
p <- p + coord_fixed()
p <- p + scale_color_gradientn(colors=c("gray90","blue","black"),
                              breaks=c(0,4,8))
p <- p + facet_wrap(~variable, nrow=2)
p <- p + theme_bw()
p <- p + guides(color = guide_colorbar(ticks = FALSE)) 
p <- p + labs(color="log count")
p <- p + theme(strip.text = element_text(face = "italic"), 
        panel.grid = element_blank(), 
        axis.title = element_blank(), 
        axis.text = element_blank(), 
        axis.ticks = element_blank())
p
```
:::

## Plotting genes with the largest p-values (least spatially variable)

```{r}
require(scran)
interesting_genes <- rownames(res$res_mtest)[order(res$res_mtest$combinedPval,decreasing=TRUE)[1:10]]

spe <- logNormCounts(spe)

Y.sub <- as.matrix(logcounts(spe)[interesting_genes,])

## Switch rows and columns and make dataframe 
df <- as.data.frame(t(Y.sub))

## Add spatial coordinates 
df$x <- spatialCoords(spe)[,1]; df$y <- spatialCoords(spe)[,2]

### Construct plot
require(reshape2)
df <- melt(df, id.vars=c("x","y"))

p <- ggplot(data=df,aes(x=x,y=y,color=value)) 
p <- p + geom_point(size=0.05) 
p <- p + coord_fixed()
p <- p + scale_color_gradientn(colors=c("gray90","blue","black"),
                              breaks=c(0,4,8))
p <- p + facet_wrap(~variable, nrow=2)
p <- p + theme_bw()
p <- p + guides(color = guide_colorbar(ticks = FALSE)) 
p <- p + labs(color="log count")
p <- p + theme(strip.text = element_text(face = "italic"), 
        panel.grid = element_blank(), 
        axis.title = element_blank(), 
        axis.text = element_blank(), 
        axis.ticks = element_blank())
p
```

# Clustering and cell type identification

## Dimension reduction

## Running PCA

```{r}
# Get the top 2000 genes.
top.hvgs <- getTopHVGs(spe, n=2000)

spe <- fixedPCA(spe, subset.row=top.hvgs)
```

## Plotting top 3 PCs 

```{r}
component <- 1

df <- data.frame(x=spatialCoords(spe)[,1],
                 y=spatialCoords(spe)[,2],
                 PC1=reducedDim(spe)[,1],
                 PC2=reducedDim(spe)[,2],
                 PC3=reducedDim(spe)[,3])

df <- df |> pivot_longer(cols=c(PC1,PC2,PC3))
p <- ggplot(data=df,aes(x=x,y=y,color=value)) + 
  geom_point(size=0.5) + 
  coord_fixed()+
  scale_color_gradient2(low="blue",mid="grey", high="red") +
  theme_bw()+
  facet_wrap(~name,nrow=1)
p
```

## Clustering

```{r}
require(igraph)
g <- buildSNNGraph(spe, use.dimred="PCA")
cluster <- igraph::cluster_walktrap(g)$membership
```

```{r}
df <- data.frame(x=spatialCoords(spe)[,1],
                 y=spatialCoords(spe)[,2],
                 cluster=as.character(cluster))
p <- ggplot(data=df,aes(x=x,y=y,color=cluster)) + 
  geom_point() + 
  theme_bw()
p
```

## Cell type identification

```{r}
colLabels(spe) <- factor(cluster)
markers <- scoreMarkers(spe)
```

## References

Ji, N., & Van Oudenaarden, A. (2012). Single molecule fluorescent in situ hybridization (smFISH) of C. elegans worms and embryos.
