# Spatial transcriptomics data analysis using R 

**Author:** Phillip Nicol

## Overview 

This repository contains materials for a workshop on spatial transcriptomics data analysis at the University of Puerto Rico. Broadly, it covers the following topics 

1.	Review of spatial transcriptomics technologies.
2.	Spatial transcriptomics data structure and storage in R.
3.	Finding spatially variable genes.
4.	Dimension reduction, clustering, and identification of marker genes.

The topics are motivated by an in-depth analysis of a real spatial transcriptomics dataset from the human brain. 

## Software prerequesites 

* `R (>= 4.0)`
* Ability to open `.qmd` files. 
* Ability to run Unix commands is preferable but not required

## Getting started 

Clone the repository, then view the presentation `UPR_ST_workshop.html`. 

## References

The analyses are based on the following online resources:

* [Introduction to the SpatialExperiment class](https://www.bioconductor.org/packages/release/bioc/vignettes/SpatialExperiment/inst/doc/SpatialExperiment.html)
* [Using scran to analyze single-cell RNA-seq data](https://bioconductor.org/packages/devel/bioc/vignettes/scran/inst/doc/scran.html)
* [Lukas Weber's BestPracticesST](https://github.com/lmweber/BestPracticesST)
